File: libyuv/src/main/java/com/theeasiestway/libyuv/YuvFrame.kt
package com.theeasiestway.libyuv

import android.media.Image
import java.nio.ByteBuffer

//
// Created by Loboda Alexey on 22.06.2020.
//

class YuvFrame {

    lateinit var y: ByteBuffer; private set
    lateinit var u: ByteBuffer; private set
    lateinit var v: ByteBuffer; private set

    var yStride: Int = 0; private set
    var uStride: Int = 0; private set
    var vStride: Int = 0; private set

    var width: Int = 0; private set
    var height: Int = 0; private set

    fun fill(image: Image) {

        for (i in 0 until 3) {
            val plane = image.planes[i]
            when(i) {
                0 -> {
                    y = plane.buffer
                    yStride = plane.rowStride
                }
                1 -> {
                    u = plane.buffer
                    uStride = plane.rowStride
                }
                2 -> {
                    v = plane.buffer
                    vStride = plane.rowStride
                }
            }
        }

        width = image.width
        height = image.height

        image.close()
    }

    fun free() {
        y = ByteBuffer.allocate(1)
        u = ByteBuffer.allocate(1)
        v = ByteBuffer.allocate(1)

        yStride = 0
        uStride = 0
        vStride = 0

        width = 0
        height = 0
    }
}

File: yuv/src/main/cpp/entities/YuvFrame.cpp
//
// Created by Loboda Alexey on 30.07.2020.
//

#include "YuvFrame.h"

YuvFrame::YuvFrame(int width, int height, int yStride, int uStride, int vStride, int ySize, int uSize, int vSize) {
    this->width = width;
    this->height = height;
    this->yStride = yStride;
    this->uStride = uStride;
    this->vStride = vStride;
    this->ySize = ySize;
    this->uSize = uSize;
    this->vSize = vSize;
    dataSize = ySize + uSize + vSize;

    y = (uint8_t*) malloc(sizeof(uint8_t) * ySize);
    u = (uint8_t*) malloc(sizeof(uint8_t) * uSize);
    v = (uint8_t*) malloc(sizeof(uint8_t) * vSize);
}

void YuvFrame::update(YuvFrame& other) {
    width = other.width;
    height = other.height;

    yStride = other.yStride;
    uStride = other.uStride;
    vStride = other.vStride;

    ySize = other.ySize;
    uSize = other.uSize;
    vSize = other.vSize;
    dataSize = other.dataSize;

    release();

    y = (uint8_t*) malloc(sizeof(uint8_t) * ySize);
    u = (uint8_t*) malloc(sizeof(uint8_t) * uSize);
    v = (uint8_t*) malloc(sizeof(uint8_t) * vSize);

    memmove(y, other.y, sizeof(uint8_t) * ySize);
    memmove(u, other.u, sizeof(uint8_t) * uSize);
    memmove(v, other.v, sizeof(uint8_t) * vSize);

    fillData();
}

void YuvFrame::fillData() {
    data = (uint8_t*) malloc(sizeof(uint8_t) * dataSize);
    memmove(data, y, sizeof(uint8_t) * ySize);
    memmove(&data[ySize], &u[0], sizeof(uint8_t) * uSize);
    memmove(&data[ySize + uSize], &v[0], sizeof(uint8_t) * vSize);
}

YuvFrame::~YuvFrame() {
    release();
}

void YuvFrame::release() {
    Frame::release();
    if (y) free(y);
    if (u) free(u);
    if (v) free(v);
    y = nullptr;
    u = nullptr;
    v = nullptr;
}